src/prognostic_equations/edmfx_closures.jl

```diff
@@ -206,78 +226,193 @@ function mixing_length(
+    most_g_m = CAP.coefficient_b_m_gryanik(params)  # Gryanik b_m
 
-    # compute the l_W - the wall constraint mixing length
-    # which imposes an upper limit on the size of eddies near the surface
-    # kz scale (surface layer)
-    if obukhov_length < 0.0 #unstable
-        l_W =
-            vkc * (ᶜz - z_sfc) /
-            max(sqrt(sfc_tke / ustar / ustar) * c_m, eps(FT)) *
-            min((1 - 100 * (ᶜz - z_sfc) / obukhov_length)^FT(0.2), 1 / vkc)
-    else # neutral or stable
-        l_W =
-            vkc * (ᶜz - z_sfc) /
-            max(sqrt(sfc_tke / ustar / ustar) * c_m, eps(FT))
+    # l_z: Geometric distance from the surface
+    l_z = ᶜz - z_sfc
+    # Ensure l_z is non-negative when ᶜz is numerically smaller than z_sfc.
+    l_z = max(l_z, FT(0))
+
+    # l_W: Wall-constrained length scale (near-surface limit, to match 
+    # Monin-Obukhov Similarity Theory in the surface layer, with Businger-Dyer 
+    # type stability functions)
+    tke_sfc_safe = max(sfc_tke, eps_FT)
+    ustar_sq_safe = max(ustar * ustar, eps_FT) # ustar^2 may vanish in certain LES setups
+
+    # Denominator of the base length scale (always positive):
+    #     c_m * √(tke_sfc / u_*²) = c_m * √(e_sfc) / u_*
+    # The value increases when u_* is small and decreases when e_sfc is small.
+    l_W_denom_factor = sqrt(tke_sfc_safe / ustar_sq_safe)
+    l_W_denom = max(c_m * l_W_denom_factor, eps_FT)
+
+    l_W = FT(0)
+    # Base length scale (neutral, but adjusted for TKE level)
+    # l_W_base = κ * z / (c_m * sqrt(e_sfc) / u_star)
+    # This can be Inf if l_W_denom is eps_FT and l_z is large.
+    # This can be 0 if l_z is 0.
+    # The expression approaches ∞ when l_W_denom ≈ eps_FT and l_z is large,
+    # and approaches 0 when l_z → 0.
+    l_W_base = vkc * l_z / l_W_denom
+
+    if obukhov_length < FT(0) # Unstable case
+        obukhov_len_safe = min(obukhov_length, -eps_FT) # Ensure L < 0
+        zeta = l_z / obukhov_len_safe # Stability parameter zeta = z/L (<0)
+
+        # Calculate MOST term (1 - b_m * zeta)
+        # Since zeta is negative, this term is > 1
+        inner_term = FT(1) - most_b_m * zeta
+
+        # Numerical safety check – by theory the value is ≥ 1.
+        inner_term_safe = max(inner_term, eps_FT)
+
+        # Unstable-regime correction factor:
+        #     (1 − b_m ζ)^(1/4) = φ_m⁻¹,
+        # where φ_m is the Businger stability function φ_m = (1 − b_m ζ)^(-1/4).
+        stability_correction = sqrt(sqrt(inner_term_safe))
+        l_W = l_W_base * stability_correction
+
+    else # Neutral or stable case
+        # Ensure L > 0 for Monin-Obukhov length
+        obukhov_len_safe_stable = max(obukhov_length, eps_FT)
+        zeta = l_z / obukhov_len_safe_stable # zeta >= 0
+
+        # Stable/neutral-regime correction after Gryanik (2020):
+        #     φ_m = 1 + a_m ζ / (1 + g_m ζ)^(2/3),
+        # a nonlinear refinement to the Businger formulation.
+        phi_m_denom_term = (FT(1) + most_g_m * zeta)
```

Comment:
Same here. If `most_g_m` and `zeta` are of type `FT` then the outcome of this line will be FT also